tests/access-control.tests.ts

```diff
@@ -0,0 +1,257 @@
+    await expect(query.hasRole(Roles.Minter, sender.address)).to.have.output(true)
+  })
+
+  it('ACCESS CONTROL - should not grant initial roles for random role', async () => {
+    const {
+      query,
+      accounts: [alice]
+    } = await setup()
+
+    await expect(query.hasRole(Roles.DefaultAdminRole, alice.address)).to.have.output(false)
+    await expect(query.hasRole(Roles.Minter, alice.address)).to.have.output(false)
+  })
+
+  it('ACCESS CONTROL - should grant role', async () => {
+    const {
+      query,
+      tx,
+      accounts: [alice]
+    } = await setup()
+
+    await tx.grantRole(Roles.Minter, alice.address)
+
+    await expect(query.hasRole(Roles.Minter, alice.address)).to.have.output(true)
+  })
+
+  it('ACCESS CONTROL - should not change old roles after grant role', async () => {
+    const {
+      query,
+      tx,
+      defaultSigner,
+      accounts: [alice]
+    } = await setup()
+
+    await tx.grantRole(Roles.Minter, alice.address)
+
+    await expect(query.hasRole(Roles.Minter, alice.address)).to.have.output(true)
+    await expect(query.hasRole(Roles.DefaultAdminRole, defaultSigner.address)).to.have.output(true)
+    await expect(query.hasRole(Roles.Minter, defaultSigner.address)).to.have.output(true)
+  })
+
+  it('ACCESS CONTROL - should revoke role', async () => {
+    const {
+      query,
+      tx,
+      accounts: [alice]
+    } = await setup()
+
+    await tx.grantRole(Roles.Minter, alice.address)
+    await tx.revokeRole(Roles.Minter, alice.address)
+
+    await expect(query.hasRole(Roles.Minter, alice.address)).to.have.output(false)
+  })
+
+  it('ACCESS CONTROL - should renounce role', async () => {
+    const {
+      query,
+      contract,
+      accounts: [alice]
+    } = await setup()
+
+    await contract.tx.grantRole(Roles.Minter, alice.address)
+    await fromSigner(contract, alice.address).tx.renounceRole(Roles.Minter, alice.address)
+
+    await expect(query.hasRole(Roles.Minter, alice.address)).to.have.output(false)
+  })
+
+  it('ACCESS CONTROL - should reject when grant/revoke not by admin role', async () => {
+    const {
+      tx,
+      contract,
+      accounts: [alice, bob]
+    } = await setup()
+
+    await tx.grantRole(Roles.Minter, bob.address)
+
+    await expect(fromSigner(contract, alice.address).tx.grantRole(Roles.Minter, alice.address)).to.eventually.be.rejected
+    await expect(fromSigner(contract, alice.address).tx.revokeRole(Roles.Minter, bob.address)).to.eventually.be.rejected
+  })
+
+  it('ACCESS CONTROL - should reject when renounce not self role', async () => {
+    const {
+      tx,
+      defaultSigner,
+      contract,
+      accounts: [alice]
+    } = await setup()
+
+    await tx.grantRole(Roles.Minter, alice.address)
+
+    await expect(fromSigner(contract, defaultSigner.address).tx.renounceRole(Roles.Minter, alice.address)).to.eventually.be.rejected
+  })
+
+  it('ACCESS CONTROL - should reject burn if no minter role', async () => {
```

Comment:
Could you also create a test where the admin will provide Minter role to Alice, and Alice calls mint please?=)

Reply:
I updated _ACCESS CONTROL - only minter role is allowed to mint_ for it